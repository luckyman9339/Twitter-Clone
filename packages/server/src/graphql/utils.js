export const decodeCursor = cursor => {
  const content = new Buffer(cursor, 'base64').toString('binary');
  const [after, order] = content.split(';');
  return { after: new Date(after), order };
};

export const encodeCursor = ({ after, order }) => {
  const input = [new Date(after).toISOString(), order].join(';');
  return Buffer.from(input).toString('base64');
};

export const buildEdgesForTweetsPagination = (tweets, order) =>
  tweets.map(tweet => {
    const type = tweet.retweetForTweetId
      ? 'RETWEET'
      : tweet.likeForTweetId
      ? 'LIKE'
      : 'TWEET';

    const node = { ...tweet, type };

    return {
      cursor: encodeCursor({
        after: tweet.created_at,
        order,
      }),
      node,
    };
  });
