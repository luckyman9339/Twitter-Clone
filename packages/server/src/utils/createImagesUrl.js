module.exports = {
  createUserAvatarUrl: email => {
    return `https://api.adorable.io/avatars/200/${email}.png`;
  },
};
