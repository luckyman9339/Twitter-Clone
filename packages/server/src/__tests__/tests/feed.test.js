import knexCleaner from 'knex-cleaner';
import { request } from '../utils';
import knex from 'db/knex';

let user = null;

beforeAll(async () => {
  await knexCleaner.clean(knex, {
    ignoreTables: ['knex_migrations', 'knex_migrations_lock'],
  });
  await knex.seed.run();
  user = (await knex.table('users').where({ username: 'test_user' }))[0];
});

it('should fetch the feed of the current user', async () => {
  const context = { user };

  const query = `
    query getUserFeed($first: Int!, $after: String) {
      feed(first: $first, after: $after) {
        edges {
          cursor
          node {
            type
            originalTweet {
              id
              content
              retweeted
              liked
              user {
                id
                name
                username
              }
            }
            contextTweet {
              id
            }
            contextUser {
              id
              name
            }
          }
        }
        pageInfo {
          hasNextPage
        }
      }
    }
  `;

  const variables = {
    first: 10,
  };

  const result = await request(query, { context, variables });
  expect(result).toMatchSnapshot();
});
