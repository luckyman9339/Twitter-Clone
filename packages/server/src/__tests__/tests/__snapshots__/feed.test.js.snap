// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should fetch the feed of the current user 1`] = `
Object {
  "data": Object {
    "feed": Object {
      "edges": Array [
        Object {
          "cursor": "MjAxOS0wNC0yNFQwMjo1MDowNC4yMzhaO2Rlc2M=",
          "node": Object {
            "__typename": "FeedNode",
            "like": Object {
              "id": "cff2eee6-f27c-4991-a4cc-08996393615b",
              "tweet": Object {
                "content": "Hello World from User Two!",
                "id": "e2a7c8d1-70c4-4e43-b7f0-a9c75d1f92ca",
                "user": Object {
                  "id": "4173dd82-7c8f-4bfa-9c8e-0158082fe07c",
                  "name": "Test User Two",
                  "username": "test_user_two",
                },
              },
              "user": Object {
                "id": "551cf7db-df68-4ad5-af6c-1935759b9291",
                "name": "Test User One",
              },
            },
            "retweet": null,
            "tweet": null,
          },
        },
        Object {
          "cursor": "MjAxOS0wNC0yNFQwMjo1MDowMy4yMzhaO2Rlc2M=",
          "node": Object {
            "__typename": "FeedNode",
            "like": null,
            "retweet": Object {
              "id": "b73c02bf-d378-4dc6-bfcd-0037d20949f4",
              "tweet": Object {
                "content": "Hello World from User Two!",
                "id": "e2a7c8d1-70c4-4e43-b7f0-a9c75d1f92ca",
                "user": Object {
                  "id": "4173dd82-7c8f-4bfa-9c8e-0158082fe07c",
                  "name": "Test User Two",
                  "username": "test_user_two",
                },
              },
              "user": Object {
                "id": "551cf7db-df68-4ad5-af6c-1935759b9291",
                "name": "Test User One",
              },
            },
            "tweet": null,
          },
        },
        Object {
          "cursor": "MjAxOS0wNC0yNFQwMjo1MDowMS4yMzhaO2Rlc2M=",
          "node": Object {
            "__typename": "FeedNode",
            "like": null,
            "retweet": null,
            "tweet": Object {
              "content": "Hello World from User One!",
              "id": "53c1c783-bebf-4f62-bd46-766c827e7d62",
              "user": Object {
                "id": "551cf7db-df68-4ad5-af6c-1935759b9291",
                "name": "Test User One",
                "username": "test_user",
              },
            },
          },
        },
      ],
      "pageInfo": Object {
        "hasNextPage": false,
      },
    },
  },
}
`;
